ARCs/arc-0022.md

```diff
@@ -0,0 +1,64 @@
+---
+arc: 22
+title: Extend arc-4 to add `read-only` annotation to methods
+status: Draft
+---
+
+# Extend arc-4 to add `read-only` annotation to methods
+
+## Summary
+
+The following document introduces a convention for creating methods (as described in [ARC-4](https://github.com/algorandfoundation/ARCs/blob/main/ARCs/arc-0004.md)) which don't mutate state.
```

Comment:
In addition to read-only, something like Solidity's pure methods also seem like a good idea. The difference between read-only and pure is that pure methods cannot read any on-chain data.

Reply:
We thought about adding pure functions but couldn't find an actual benefit for that. Any ideas?

Reply:
Could someone please explain to me what exactly is the benefit of a readonly flag or a pure flag? I am asking because I feel like I am missing something. AFAIK these ABIs are post-hoc descriptive entities and not used for driving compilation (eg: the pure flag in solidity), so how are you expecting these flags to be used and by whom?

Reply:
I think @jasonpaulos summed it up in the first comment, but I'll elaborate. If a function is marked "read-only" then an off-chain evaluator could confidently run an avm evaluator against block chain state, and get a correct answer.  This is like a view function in Ethereum.  You're not looking to send transaction to the block chain with the attendant costs and latency - you just want to run the code to get an answer where the code is really just interpreting the contracts internal state.

A pure function must the same, except running the code is even easier - it won't look at blockchain state.  So your IDE could have an AVM interpreter, or perhaps your unittest framework. And it wouldn't even need access to blockchain state to test your functions.

Both are useful for dapps that are trying to display information or build txns using code from the contract rather than interpret bytes of state on their own.

Reply:
Maybe I will leave this until tomorrow as my brain must have failed or something. I totally absolutely don't get it. 

So in scenario 2 above I completely ignore if someone has set a readonly flag, send my txn to the 'sandbox endpoint' or whatever it is, and if there's a state access it needs to fail anyway. Why not just have metadata in the contract that it's readonly and enforce it in the AVM? Anyway, I guess I'll leave the convo for now, I probably need a break.

Reply:
I'm having this discussion with @barnjamin right now, so I want to point out that the key benefit of an explicit "read-only" flag is that decision to do number 1 or number 2 from Jason's example can be automated.  dapp writer writes a method invocation and he gets an answer. But if it's read-only, he doesn't pay for the answer. He doesn't consult the description, read it, and decide "Oh, since it says read-only in the prose, I can invoke it the special cheap way". That happens for him.

Reply:
Perhaps there are also benefits to making it a property of the app call instead. Perhaps an abi method can be read only or not depending on a switch property of the txn with the app called in a read only mode. 

By the way what would be the behavkour  for group transactions including these?